mlx/linalg.cpp

```diff
@@ -172,4 +173,31 @@ array norm(
   return matrix_norm(a, ord, ax, keepdims, s);
 }
 
+std::pair<array, array> qr(const array& a, StreamOrDevice s /* = {} */) {
+  if (a.dtype() != float32) {
+    std::ostringstream msg;
+    msg << "[linalg::qr] Arrays must type float32. Received array "
+        << "with type " << a.dtype() << ".";
+    throw std::invalid_argument(msg.str());
+  }
+  if (a.ndim() < 2) {
+    std::ostringstream msg;
+    msg << "[linalg::qr] Arrays must have >= 2 dimensions. Received array "
+           "with "
+        << a.ndim() << " dimensions.";
+    throw std::invalid_argument(msg.str());
+  }
+  if (a.shape(-1) != a.shape(-2)) {
+    throw std::invalid_argument(
+        "[linalg::qr] Support for non-square matrices NYI.");
```

Comment:
@nullhook @awni out of curiosity: what's the scope of work to support non-square?  It seems like the lapack funcs `sgeqrf` and `sorgqr` used here already support it. 

IIUC, for a non-square matrix A with M rows and N columns process is:
1. Call `sgeqrf` to compute the QR factorization A = Q*R, where Q is now an MxM  matrix, and R is an MxN upper trapezoidal matrix.
2. Call`sorgqr` to generate the MxM matrix Q from the reflectors (similar to current impl)

So I supposed the scope is to define some helper logic to calculate min(M, N) for estimating size of the trap matrix R; and consequently memory allocations for the various arrays.  (And perhaps a few other related details around the for-loops halting conditions etc)

Im curious if that's the right line of thinking? OR do you see anything major I could be missing?

Reply:
That sounds about right to me